Introduce a normalization chapter (#2266)
rust-lang/rustc-dev-guide: src/normalization.md

If we were to normalize it to `for<'a> fn(?y)` and register a goal to normalize `for<'a> <?x as Trait<'a>>::Assoc -> ?y`, this would result in errors in cases where `<?x as Trait<'a>>::Assoc` normalized to `&'a u32`. This is because the inference variable `?y` would be in a lower [universe][universes] than the placeholders made when instantiating the `for<'a>` binder.

Leaving the alias unnormalized would also be wrong as the old solver expects all aliases to be rigid. This was a soundness bug before the new solver was stabilized in coherence: [relating projection substs is unsound during coherence](https://github.com/rust-lang/rust/issues/102048).

Ultimately this means that it is not always possible to ensure all aliases inside of a value are rigid.

[universes]: https://rustc-dev-guide.rust-lang.org/borrow_check/region_inference/placeholders_and_universes.html#what-is-a-universe
[deeply_normalize]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_trait_selection/traits/normalize/trait.NormalizeExt.html#tymethod.deeply_normalize

## Handling uses of diverging aliases

> > NOTE: This section should talk about how we handles uses of diverging aliases. writeback normalizes everything explicitly, lots of places are adhoc and accidentally error or don't manage to error even though they ought to, const aliases are a bit more "principled" about checking for termination.

Author: Imagine learning that writeback does not infact normalize everything explicitly. what a concept

[tykind_alias]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_type_ir/enum.TyKind.html#variant.Alias
[aliaskind]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_type_ir/enum.AliasTyKind.html

### Rigid, Ambiguous and Unnormalized Aliases

Aliases can either be "rigid", "ambiguous", or simply unnormalized.

We consider types to be rigid if their "shape" isn't going to change, for example `Box` is rigid as no amount of normalization can turn a `Box` into a `u32`, whereas `<vec::IntoIter<u32> as Iterator>::Item` is not rigid as it can be normalized to `u32`.

Aliases are rigid when we know that we will never be able to normalize them further. A concrete example of a *rigid* alias would be `<T as Iterator>::Item` in an environment where there is no `T: Iterator<Item = ...>` bound, only a `T: Iterator` bound:
```rust
fn foo<T: Iterator>() {
    // This alias is *rigid*
    let _: <T as Iterator>::Item;
}

fn bar<T: Iterator<Item = u32>>() {
    // This alias is *not* rigid as it can be normalized to `u32`
    let _: <T as Iterator>::Item;
}
```

> This definition also implies that illformed aliases can be considered rigid. For example the type `<u32 as Iterator>::Item` is a rigid alias as it can never be normalized as there is no `u32: Iterator` impl (ignoring `feature(trivial_bounds)`).

Reviewer: that doesn't seem right. `<u32 as Iterator>::Item` is not rigid in the new solver and iirc the old solver also uses `normalizes_to_error` in these cases https://github.com/rust-lang/rust/blob/efcbb94dccbe0fcc1076f031f88332db4a6a2a52/compiler/rustc_trait_selection/src/traits/project.rs#L459

Author: oh neat okay :3

When an alias can't yet be normalized but may eventually wind up normalizable we consider it to be an "ambiguous" alias. This can occur when an alias contains inference variables which prevent being able to determine how the trait is implemented:
```rust
fn foo<T: Iterator, U: Iterator>() {
    // This alias is considered to be "ambiguous"
    let _: <_ as Iterator>::Item;
}
```

The reason we call them "ambiguous" aliases is because its *ambiguous* whether this is a rigid alias or not.

> The source of the `_: Iterator` trait impl is *ambiguous* (i.e. unknown), it could be some `impl Iterator for u32` or it could be some `T: Iterator` trait bound, we don't know yet. Depending on how `_: Iterator` holds the alias could be an unnormalized alias or it could be a rigid alias; it's *ambiguous* what kind of alias this is.

Reviewer: ```suggestion
The source of the `_: Iterator` trait impl is *ambiguous* (i.e. unknown), it could be some `impl Iterator for u32` or it could be some `T: Iterator` trait bound, we don't know yet. Depending on why `_: Iterator` holds the alias could be an unnormalized alias or it could be a rigid alias; it's *ambiguous* what kind of alias this is.
```

Finally, an alias can just be unnormalized, `<Vec<u32> as IntoIterator>::Iter` is an unnormalized alias as it can already be normalized to `std::vec::IntoIter<u32>`, it just hasn't been done yet.

---

> It is worth noting that Free and Inherent aliases cannot be rigid or ambiguous as   naming them also implies having resolved the definition of the alias, which specifies the underlying type of the alias:

Reviewer: ```suggestion
It is worth noting that Free and Inherent aliases cannot be rigid or ambiguous as  naming them also implies having resolved the definition of the alias, which specifies the underlying type of the alias:
```

```rust
#![feature(inherent_associated_types)]

trait Trait {}

impl Trait for u32 {}
impl Trait for u64 {}

struct Foo<T>(T);

impl<T: Trait> Foo<T> {
    type Inherent = u32;
}

fn ret_alias<T: Trait>(_: T) -> Foo<T>::Inherent {
    10_u32
}

fn main() {
    // `t` has type `?x`
    let mut t = Default::default();
    // `ret_alias` returns a value of type `Foo<?x>::Inherent`
    t = ret_alias(t);
    
    // we have equated `?x` with `Foo<?x>::Inherent` this can
    // only succeed if we're able to normalize inherent aliases
    // with ambiguous where clauses on the impl.
    //
    // Or, put another way, it is *not* an ambiguous alias.
}

> ```

Reviewer: What are you trying to show with this example? It feels confusing to me (and too "large" for showing that inherent aliases are always normalizeable)

Author: Yeah idk removed it

### Diverging Aliases

> An alias is considered to "diverge" if its definition does not actually specify an underlying non-alias type to normalize to. On the flip side, we consider an alias to "terminate" if it does have an underlying value it can be normalized to (i.e. it does not diverge). A concrete example of diverging aliases:

Author: Looking at some private convo chat logs I definitely see a lot of talking about "alias terminates", atleast enough to justify explaining the term here as people will probably encounter it in the wild.

> An alias is considered to "diverge" if its definition does not actually specify an underlying non-alias type to normalize to. On the flip side, we consider an alias to "terminate" if it does have an underlying value it can be normalized to (i.e. it does not diverge). A concrete example of diverging aliases:

Author: Thinking about it more, we only really use "terminates" terminology like twice, so it's probably More confusing to keep it here instead of using diverging everywhere.

```rust
type Diverges = Diverges;

trait Trait {
    type DivergingAssoc;
}
impl Trait for () {
    type DivergingAssoc = <() as Trait>::DivergingAssoc;
}
```
In this example both `Diverges` and `DivergingAssoc` are "trivial" cases of diverging type aliases where they have been defined as being equal to themselves. There is no underlying type that `Diverges` can ever be normalized to.

We generally try to error when diverging aliases are defined, but this is entirely a "best effort" check. In the previous example the definitions are "simple enough" to be detected and so errors are emitted. However, in more complex cases, or cases where only some instantiations of generic parameters would result in a diverging alias, we don't emit an error:
```rust
trait Trait {
    type DivergingAssoc<U: Trait>;
}
impl<T: ?Sized> Trait for T {
    // This alias always diverges but we don't emit an error because
    // the compiler can't "see" that.
    type DivergingAssoc<U: Trait> = <U as Trait>::DivergingAssoc<U>;
}
```

> Ultimately this means that we have no guarantee that aliases in the type system terminate. As aliases may only diverge for some specific arguments to generic parameters, it also means that only fully concrete aliases can be checked to terminate, which means that codegen/const evaluation is the only time that aliases can reliably be checked to terminate:

Reviewer: ```suggestion
Ultimately this means that we have no guarantee that aliases in the type system don't diverge. As aliases may only diverge for some specific arguments to generic parameters, it also means that we only know whether an alias diverges once it is fully concrete. This means that codegen/const evaluation also has to handle diverging aliases can reliably be checked to terminate:
```
I personally dislike the perspective "only codegen can check". It's more that "codegen has to check for them *as well*"

Author: I'm not really sure I understand the distinction :thinking: Codegen has to check for them because *only* codegen can check some aliases.

Author: I updated the wording anyway :thinking:

As such there is no `ConstKind::Alias(AliasCtKind::Projection/Inherent/Free, _)`, instead we only have `ConstKind::Unevaluated` which is used for representing anonymous constants.

```rust
fn foo<const N: usize>() {}

const FREE_CONST: usize = 1 + 1;

fn bar() {
    foo::<{ FREE_CONST }>();
    // The const arg is represented with some anonymous constant:
    // ```pseudo-rust
    // const ANON: usize = FREE_CONST; 
    // foo::<ConstKind::Unevaluated(DefId(ANON), [])>();
    // ```
}
```

This is likely to change as const generics functionality is improved, for example `feature(associated_const_equality)` and `feature(min_generic_const_args)` both require handling const aliases similarly to types (without an anonymous constant wrapping all const args).

## Normalization

There are two main "parts" to how normalization exists in the compiler:
- The core logic responsible for turning an alias into its underlying type
- The various entry points to the core normalization logic that expose real APIs

### Core normalization logic

> Normalizing aliases is a little bit more nuanced than replacing the alias with whatever it was defines as being equal to in its definition; the result of normalizing an alias should either be a rigid type or an inference variable (which will later be inferred to a rigid type). To accomplish this we do two things:

Reviewer: ```suggestion
Normalizing aliases is a little bit more nuanced than replacing the alias with whatever it is defined as being equal to in its definition; the result of normalizing an alias should either be a rigid type or an inference variable (which will later be inferred to a rigid type). To accomplish this we do two things:
```

First, when normalizing an ambiguous alias it is normalized to an inference variable instead of leaving it as-is, this has two main effects: 

> - Even though an inference variable is not a rigid type, it will always wind up inferred *to* a rigid type so we ensure that the result of normalization will not need to be normalized again

Reviewer: shallow vs deep normalization. idk how to best introduce this, but it only guarantees that the "outer type constructor" is rigid.

you could normalize to `fn(<&() as Trait>::Assoc>)` after all which contains a normalizeable alias.

I think before explaining *how* normalization works, it may make sense to introduce "structural normalization" and "deep normalization"

Author: Yeah it feels like its missing a section specifically calling out shallow vs deep norm :thinking: It's awkward because the old solver kinda likes to pretend that normalization is deep, yet `deeply_normalize` exists and is arguably just as "deep" as shallow norm in the old solver.

Even in the new solver "shallow" normalization is largely deep norm with the exception of aliases in binders </3

Author: Or well, deeply_normalize *is* arguably more deep in the sense that you never have to worry about encountering non-rigid aliases inside of the term anymore. But that feels so subtle given it only *really* matters in the edge case of hr aliases :sweat_smile:

There is not much of a distinction between the three methods of normalizing with `InferCtxt`. The primary reason for there to exist all three is as a transitionary step between the old and new solver, because in the new solver there is a distinction between deeply normalizing and structurally normalizing, and the old solver behaviour no longer exists. 

In the new solver:
- `infcx.at.normalize`: A no-op, `deeply_normalize` or `structurally_normalize` should be used instead. This method only exists for the old solver.
- `deeply_normalize`: Folds over the provided value normalizing all aliases, erroring on ambiguous aliases[^1].
- `structurally_normalize`: Normalizes an alias to a rigid type or an inference variable, does nothing if the type is not an alias.

When the new solver is stabilized the `infcx.at.normalize` function will be removed and everything will have been migrated to either deeply normalize or structurally normalize. In cases where normalization is required in both old and new solvers, the `structurally_normalize` function is often the right choice. Otherwise, if normalization is only needed in the old solver then `infcx.at.normalize` is often the right choice.

> `structurally_normalize` should be thought of as only normalizing the "outer layer" of a type to a rigid type or inference variable. Any types inside of the normalized form may still be unnormalized (even if an alias *was* actually normalized[^2]). `structurally_normalize` is named such as it is typically called before looking at the "structure" of a type (e.g. by matching on `TyKind`/`ConstKind`).

Reviewer: i think you probably should move this up without yet talking about impl details.

shallow vs deep normalization is an important distinction

[^2]: Ambiguous aliases inside of binders cannot be normalized to inference variables, this will be covered more later.

[normalize]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_trait_selection/infer/at/struct.At.html#method.normalize
[deeply_normalize]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_trait_selection/traits/normalize/trait.NormalizeExt.html#tymethod.deeply_normalize
[structurally_normalize]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_trait_selection/traits/trait.StructurallyNormalizeExt.html#tymethod.structurally_normalize_ty
[infcx]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_trait_selection/infer/struct.InferCtxt.html
[fcx]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir_typeck/fn_ctxt/struct.FnCtxt.html
[ocx]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_trait_selection/traits/struct.ObligationCtxt.html
[structurally_resolve]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir_typeck/fn_ctxt/struct.FnCtxt.html#method.structurally_resolve_type

#### `tcx.normalize_erasing_regions`

> [`normalize_erasing_regions`][norm_erasing_regions] is generally used by parts of the compiler that are not doing type system analysis' as this normalization entry point does not handle inference variables, lifetimes, or any diagnostics. Lints and codegen make heavy use of this entry point as they typically are working with fully inferred aliases that can be assumed to be well formed (or at least, are not responsible for erroring on). 

Reviewer: ```suggestion
[`normalize_erasing_regions`][norm_erasing_regions] is generally used by parts of the compiler that are not doing type system analysis. This normalization entry point does not handle inference variables, lifetimes, or any diagnostics. Lints and codegen make heavy use of this entry point as they typically are working with fully inferred aliases that can be assumed to be well formed (or at least, are not responsible for erroring on). 
```

---

There were a few main issues with the old solver's approach to normalization that motivated changing things in the new solver:

### Missing normalization calls

It was a frequent occurrence that normalization calls would be missing, resulting in passing unnormalized types to APIs expecting everything to already be normalized. Treating ambiguous or unnormalized aliases as rigid would result in all sorts of weird errors from aliases not being considered equal to one another, or surprising inference guidance from equating unnormalized aliases' generic arguments.

### Normalizing parameter environments

Another problem was that it was not possible to normalize `ParamEnv`s correctly in the old solver as normalization itself would expect a normalized `ParamEnv` in order to give correct results. See the chapter on `ParamEnv`s for more information: [`Typing/ParamEnv`s: Normalizing all bounds](./typing_parameter_envs.md#normalizing-all-bounds)

### Ambiguous aliases in higher ranked types

Given a type such as `for<'a> fn(<?x as Trait<'a>::Assoc>)`, it is not possible to correctly handle this with the old solver's approach to normalization.

> If we were to normalize it to `for<'a> fn(?y)` and register a goal to normalize `for<'a> <?x as Trait<'a>>::Assoc -> ?y`, this would result in errors in cases where `<?x as Trait<'a>>::Assoc` normalized to `&'a u32`. This is because the inference variable `?y` would be in a lower [universe][universes] than the placeholders made when instantiating the `for<'a>` binder.

Reviewer: ```suggestion
If we were to normalize it to `for<'a> fn(?y)` and register a goal to normalize `for<'a> <?x as Trait<'a>>::Assoc -> ?y`, this would result in errors in cases where `<?x as Trait<'a>>::Assoc` normalized to `&'a u32`. The inference variable `?y` would be in a lower [universe][universes] than the placeholders made when instantiating the `for<'a>` binder.
```
